File: 40.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        if not candidates: return []
        candidates.sort()
        res = []
        self.dfs(candidates, target, 0, [], res)
        return res
    def dfs(self, candidates, target, index, path, res):
        if target<0:
            return
        if target ==0:
            res.append(path)
            return
        for i in range(index, len(candidates)):
            if i != index and candidates[i] == candidates[i-1]:
                continue
            self.dfs(candidates, target - candidates[i], i + 1, path + [candidates[i]], res)

File: 40.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        if (candidates.size() == 0){
            return {};
        }
        std::sort(candidates.begin(), candidates.end());
        vector<vector<int>> res;
        vector<int> tmp;
        dfs(candidates, target, 0, res, tmp);
        return res;
    }
    void dfs(vector<int>& candidates, int target, int index, vector<vector<int>>& res, vector<int>& tmp){
        if (target < 0){
            return;
        }
        if (target == 0){
            res.push_back(tmp);
            return;
        }
        for (int i = index; i < candidates.size();i++){
            if (i != index and candidates[i] == candidates[i-1]){
                continue;
            }
            tmp.push_back(candidates[i]);
            dfs(candidates, target - candidates[i], i+1, res, tmp);
            tmp.pop_back();
        }
    }
};
